### PR TITLE
TFLite models on different backends (tests and improvements)

### PR DESCRIPTION
**Pull request by dkurt**
### Pull Request Readiness Checklist

* MaxUnpooling with OpenVINO
* Fully connected with transposed inputs/weights with OpenVINO
* Enable backends tests for TFLite (related to https://github.com/opencv/opencv/issues/23992#issuecomment-1640691722)
* Increase existing tests thresholds

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom

Xbuild_image:Custom=ubuntu-openvino-2021.4.2:20.04
build_image:Custom=ubuntu-openvino-2022.1.0:20.04

test_modules:Custom=dnn,python2,python3,java,gapi,video

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=ON
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

**Comment by dkurt** [comment]
@asmorkalov, do we still have an option to run CI with OpenVINO setup somehow?

**Comment by opencv-alalek** [comment]
> option to run CI with OpenVINO setup

Legacy CI:
- weekly: https://pullrequest.opencv.org/buildbot/builders/4_x_openvino-skl-lin64 (openvino-2021.4.2)
- CI configuration for PR has been added

**Comment by asmorkalov** [comment]
I see the following failure with the patch:
```
./bin/opencv_test_dnn --gtest_filter="DNNTestNetwork.FastNeuralStyle_eccv16*" 
...
[==========] Running 3 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 3 tests from DNNTestNetwork
[ RUN      ] DNNTestNetwork.FastNeuralStyle_eccv16/0, where GetParam() = NGRAPH/CPU
/mnt/projects/Projects/OpenCV/opencv-master/modules/dnn/test/test_common.impl.hpp:79: Failure
Expected: (normInf) <= (lInf), actual: 0.00276947 vs 0.002
First run  |ref| = 148.53863525390625
/mnt/projects/Projects/OpenCV/opencv-master/modules/dnn/test/test_common.impl.hpp:79: Failure
Expected: (normInf) <= (lInf), actual: 0.00216675 vs 0.002
Second run  |ref| = 148.02413940429688
[  FAILED  ] DNNTestNetwork.FastNeuralStyle_eccv16/0, where GetParam() = NGRAPH/CPU (1346 ms)
[ RUN      ] DNNTestNetwork.FastNeuralStyle_eccv16/1, where GetParam() = OCV/OCL
[ WARN:0@1.866] global ocl4dnn_conv_spatial.cpp:1923 loadTunedConfig OpenCV(ocl4dnn): consider to specify kernel configuration cache directory through OPENCV_OCL4DNN_CONFIG_PATH parameter.
Could not open file: /tmp/dep-02b26e.d
OpenCL program build log: dnn/dummy
Status -11: CL_BUILD_PROGRAM_FAILURE
-cl-no-subgroup-ifp
Error in processing command line: Don't understand command line argument "-cl-no-subgroup-ifp"!
/mnt/projects/Projects/OpenCV/opencv-master/modules/dnn/test/test_common.impl.hpp:79: Failure
Expected: (normInf) <= (lInf), actual: 0.00235748 vs 0.002
Second run  |ref| = 148.02413940429688
[  FAILED  ] DNNTestNetwork.FastNeuralStyle_eccv16/1, where GetParam() = OCV/OCL (1065 ms)
[ RUN      ] DNNTestNetwork.FastNeuralStyle_eccv16/2, where GetParam() = OCV/OCL_FP16
[       OK ] DNNTestNetwork.FastNeuralStyle_eccv16/2 (707 ms)
[----------] 3 tests from DNNTestNetwork (3118 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 1 test case ran. (3118 ms total)
[  PASSED  ] 1 test.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] DNNTestNetwork.FastNeuralStyle_eccv16/0, where GetParam() = NGRAPH/CPU
[  FAILED  ] DNNTestNetwork.FastNeuralStyle_eccv16/1, where GetParam() = OCV/OCL
```
Not sure if it's relevant. Will check 4.x and return back with details.

**Comment by asmorkalov** [comment]
The issue is not relevant to the PR and exists in 4.x too. OpenVINO version: openvino_toolkit_ubuntu18_2022.1.1.7030.39aba80957e

File: modules/dnn/src/layers/max_unpooling_layer.cpp [hunk 1]
```diff
@@ -181,6 +183,44 @@ class MaxUnpoolLayerImpl CV_FINAL : public MaxUnpoolLayer
...
+#ifdef HAVE_DNN_NGRAPH
+    virtual Ptr<BackendNode> initNgraph(const std::vector<Ptr<BackendWrapper> >& inputs,
+                                        const std::vector<Ptr<BackendNode> >& nodes) CV_OVERRIDE
+    {
+        auto features = nodes[0].dynamicCast<InfEngineNgraphNode>()->node;
+        auto indices = nodes[1].dynamicCast<InfEngineNgraphNode>()->node;
+
+        auto inpShape = features->get_shape();
+        std::vector<int32_t> outShape(inpShape.begin(), inpShape.end());
+        outShape[2] = (outShape[2] - 1) * poolStride.height + poolKernel.height - 2 * poolPad.height;
```
Comment (asmorkalov) [inline modules/dnn/src/layers/max_unpooling_layer.cpp:196]: I propose to add assert to check size of `outShape` before access

File: modules/dnn/src/layers/max_unpooling_layer.cpp [hunk 2]
```diff
@@ -41,6 +42,7 @@ class MaxUnpoolLayerImpl CV_FINAL : public MaxUnpoolLayer
     {
         return backendId == DNN_BACKEND_OPENCV ||
                backendId == DNN_BACKEND_CUDA ||
+               backendId == DNN_BACKEND_INFERENCE_ENGINE_NGRAPH||
```
Comment (asmorkalov) [inline modules/dnn/src/layers/max_unpooling_layer.cpp:45]: Formatting.

File: modules/dnn/src/layers/max_unpooling_layer.cpp (hunk 1 above)
Comment (dkurt) [reply modules/dnn/src/layers/max_unpooling_layer.cpp]: Replaced to manual call of `getMemoryShapes`

**Comment by dkurt** [comment]
```
[  PASSED  ] 13143 tests.
[  FAILED  ] 3 tests, listed below:
[  FAILED  ] Test_Caffe_nets.FasterRCNN_vgg16/2, where GetParam() = NGRAPH/OCL_FP16
[  FAILED  ] Test_Darknet_nets.YOLOv4/2, where GetParam() = NGRAPH/OCL_FP16
[  FAILED  ] Test_ONNX_layers.LSTM_init_h0_c0/2, where GetParam() = NGRAPH/OCL_FP16
```

**Comment by asmorkalov** [comment]
@dkurt could you rebase the patch and fix conflicts.

**Comment by dkurt** [comment]
@asmorkalov, thanks! I will fix it.

File: modules/dnn/test/test_halide_layers.cpp [hunk 3]
```diff
@@ -361,22 +361,9 @@ TEST_P(MaxPooling, Accuracy)
...
-#if defined(INF_ENGINE_RELEASE) && INF_ENGINE_VER_MAJOR_EQ(2018050000)
-    if (backendId == DNN_BACKEND_INFERENCE_ENGINE_NN_BUILDER_2019 && targetId == DNN_TARGET_MYRIAD
-            && (kernel == Size(2, 2) || kernel == Size(3, 2))
-            && stride == Size(1, 1) && (pad == Size(0, 0) || pad == Size(0, 1))
-    )
-        applyTestTag(CV_TEST_TAG_DNN_SKIP_IE_MYRIAD, CV_TEST_TAG_DNN_SKIP_IE_VERSION);
-#endif
+    // https://github.com/openvinotoolkit/openvino/issues/18731
+    if (backendId == DNN_BACKEND_INFERENCE_ENGINE_NGRAPH && stride != Size(1, 1))
+        applyTestTag(CV_TEST_TAG_DNN_SKIP_IE_NGRAPH);
```
Comment (dkurt) [inline modules/dnn/test/test_halide_layers.cpp:366]: @asmorkalov, please do not merge, I'm going to workaround it as suggested in the ticket.

Comment (dkurt) [reply modules/dnn/test/test_halide_layers.cpp:366]: Please ignore, think must complicate this PR too much. Let's just skip for now

File: modules/dnn/test/test_tflite_importer.cpp [hunk 4]
```diff
@@ -101,7 +152,15 @@ TEST(Test_TFLite, max_unpooling)
     net.setInput(input);
 
     std::vector<std::vector<Mat> > outs;
-    net.forward(outs, {"p_re_lu_1", "max_pooling_with_argmax2d", "conv2d_86", "max_unpooling2d_2"});
+    if (backend == DNN_BACKEND_INFERENCE_ENGINE_NGRAPH) {
+        // TODO: seems like a bug with a retrieving intermediate tensors
```
[the remark this answers is not in the document]
Comment (dkurt) [reply modules/dnn/test/test_tflite_importer.cpp:156]: https://github.com/opencv/opencv/issues/24102

**Comment by dkurt** [comment]
> @asmorkalov, I also waiting for https://pullrequest.opencv.org/buildbot/builders/precommit_custom_linux/builds/100310 but it runs too long on the latest commit (previous were fine). So need to check that problem is not in the changes from the latest commit.

@asmorkalov, seems like Custom worker got stuck. The job is always "queued".

**Comment by asmorkalov** [comment]
The last version passes all tests with l_openvino_toolkit_ubuntu18_2022.1.1!

**Review by asmorkalov — APPROVED** [review]
:+1:

File: modules/dnn/src/layers/pooling_layer.cpp [hunk 5]
```diff
@@ -613,9 +613,17 @@ class PoolingLayerImpl CV_FINAL : public PoolingLayer
             return Ptr<BackendNode>(new InfEngineNgraphNode(reduce_sum));
         }
         else if (type == MAX) {
-            auto max_pool = std::make_shared<ngraph::op::v1::MaxPool>(ieInpNode, ngraph::Strides(strides),
-                            ngraph::Shape(pads_begin), ngraph::Shape(pads_end), ngraph::Shape(kernel_size),
-                            rounding_type, pad_type);
+            std::shared_ptr<ngraph::Node> max_pool;
+            if (computeMaxIdx) {
+                std::vector<size_t> dilations(kernel_size.size(), 1);
+                max_pool = std::make_shared<ngraph::op::v8::MaxPool>(ieInpNode, ngraph::Strides(strides), ngraph::Strides(dilations),
```
Comment (opencv-alalek) [inline modules/dnn/src/layers/pooling_layer.cpp:619]: There are broken weekly builds:

```
/build/4_x_openvino-opencl-skl-lin64/opencv/modules/dnn/src/layers/pooling_layer.cpp: In member function 'virtual cv::Ptr<cv::dnn::dnn4_v20230620::BackendNode> cv::dnn::PoolingLayerImpl::initNgraph(const std::vector<cv::Ptr<cv::dnn::dnn4_v20230620::BackendWrapper> >&, const std::vector<cv::Ptr<cv::dnn::dnn4_v20230620::BackendNode> >&)':
/build/4_x_openvino-opencl-skl-lin64/opencv/modules/dnn/src/layers/pooling_layer.cpp:619:57: error: 'v8' is not a member of 'ngraph::op'; did you mean 'v7'?
  619 |                 max_pool = std::make_shared<ngraph::op::v8::MaxPool>(ieInpNode, ngraph::Strides(strides), ngraph::Strides(dilations),
      |                                                         ^~
      |                                                         v7
```

https://pullrequest.opencv.org/buildbot/builders/4_x_openvino-opencl-skl-lin64/builds/100029

We still use OpenVINO 2021.4

Comment (dkurt) [reply modules/dnn/src/layers/pooling_layer.cpp:619]: Thanks, I will add a check

**Pull request merged**